git-traverse/src/commit.rs

```diff
@@ -109,6 +167,63 @@ pub mod ancestors {
             res.map(Ok)
         }
     }
+
+    mod predicates {
+        use super::*;
+
+        /// `FnMut`-like trait used for pruning commit subgraphs from the ancestors iterator.
+        ///
+        /// This trait only exists because:
+        /// 1. Predicate usage should be optional.
+        ///    The user should opt-in to using predicates via an alternate constructor or via the
+        ///    builder pattern.
+        /// 2. The default case (include all ancestors) should not increase the size of the
+        ///    [`Ancestors`] struct, nor should it require extra branches at runtime.
+        ///    So don't use boxed closures, `Option<impl FnMut...>`, or `fn(&oid) -> bool`.
+        ///    Instead, use a zero-sized type ([`AlwaysTrue`]) to handle the default case.
+        /// 3. Directly implementing `FnMut` is not yet supported in stable Rust.
+        /// 4. I could not get the compiler to accept [`Ancestors::new`] using a closure or static
+        ///    function as the default predicate. I don't think this is possible without using
+        ///    something like `typeof!(always_true)` (where `always_true` is a static function) as a
+        ///    type parameter.
+        pub trait OidPredicate {
+            /// Indicate whether the given commit and its parents should be visited by the
+            /// [`Ancestors`] iterator.
+            ///
+            /// Even if this method returns `false`, the commit's ancestors may still be visited if
+            /// they are reachable from another commit for which this method returns `true`.
+            /// If you still want to exclude those ancestors, then this method should also return
+            /// `false` for those ancestor commits.
+            fn call_mut(&mut self, oid: &oid) -> bool;
+        }
+
+        /// Default behavior to include all ancestor commits in the [`Ancestors`] iterator.
+        pub struct AlwaysTrue;
+
+        impl OidPredicate for AlwaysTrue {
+            #[inline(always)]
+            fn call_mut(&mut self, _: &oid) -> bool {
+                true
+            }
+        }
+
+        pub struct FnMutWrapper<P>
+        where
+            P: FnMut(&oid) -> bool,
+        {
+            pub predicate: P,
+        }
```

Comment:
Regarding point 2 of `OidPredicate`, isn't this effectively a function pointer? There's still a layer of indirection with this, but instead of using a fn pointer you're instead using a struct. New type pattern instead, maybe?

Reply:
I didn't verify this, but I assumed that `P` and `FnMutWrapper` would be a zero-sized types with static dispatch when `AlwaysTrue` was used.

I think newtype would give the same results.

Reply:
Looks like you're correct! https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9b7dc00fe4aaf269357b4b3c37f5091d